Apptainer 1.1.0 doesn't work with lhcathome atlas unlike apptainer 1.0.3

### Version of Apptainer

What version of Apptainer (or Singularity) are you using? Run


apptainer-1.1.0~rc.1-1
### Expected behavior

Lhc@home atlas native works normally just like on apptainer version 1.0.3-1

[Новый файл1.txt](https://github.com/apptainer/apptainer/files/9308373/1.txt)


### Actual behavior
[Новый файл.txt](https://github.com/apptainer/apptainer/files/9308372/default.txt)


<!-- if this is a feature request, you can ignore this next part -->

### Steps to reproduce this behavior

Install apptainer-1.1.0~rc.1-1.el8.x86_64.rpm

### What OS/distro are you running

```NAME="Rocky Linux"
VERSION="8.6 (Green Obsidian)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.6"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8.6 (Green Obsidian)"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:rocky:rocky:8:GA"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky Linux"
ROCKY_SUPPORT_PRODUCT_VERSION="8"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8"


```

### How did you install Apptainer
From EPEL Testing x86_64
https://centos.pkgs.org/8/epel-testing-x86_64/apptainer-1.1.0~rc.1-1.el8.x86_64.rpm.html

Starting with centos stream 8 shouldn't make any difference. 

I figured out what was causing the 'Computation error' on my machine by looking in /var/lib/boinc/slots/0/stderr.txt.  My guess about a GPU was all wrong.  The problem was that I had enabled cgroups v2 on my machine.  I disabled that in /etc/default/grub, re-made the grub config, and rebooted, and now jobs are running on my VM.  I don't even see it using singularity at all, it is using /cvmfs/grid.cern.ch/vc/containers/runc.  So why is it trying to use apptainer on your machine?

I tried disabling unprivileged user namespaces with `sysctl user.max_user_namespaces=0` and then the jobs immediately exited with `[ERROR] max_user_namespaces in /proc/sys/user/ is set to 0`.  So that must be enabled on your machine and apptainer should work without apptainer-suid.  Are you are somehow running some different types of job based on singularity instead of runc?

It will use apptainer/singularity for atlas if you disable theory and 
enable atlas at https://lhcathome.cern.ch/lhcathome/prefs.php?subset=project

12.08.2022 22:05:44, DrDaveD пишет:
>
> Starting with centos stream 8 shouldn't make any difference.
>
> I figured out what was causing the 'Computation error' on my machine 
> by looking in /var/lib/boinc/slots/0/stderr.txt. My guess about a GPU 
> was all wrong. The problem was that I had enabled cgroups v2 on my 
> machine. I disabled that in /etc/default/grub, re-made the grub 
> config, and rebooted, and now jobs are running on my VM. I don't even 
> see it using singularity at all, it is using 
> /cvmfs/grid.cern.ch/vc/containers/runc. So why is it trying to use 
> apptainer on your machine?
>
> I tried disabling unprivileged user namespaces with |sysctl 
> user.max_user_namespaces=0| and then the jobs immediately exited with 
> |[ERROR] max_user_namespaces in /proc/sys/user/ is set to 0|. So that 
> must be enabled on your machine and apptainer should work without 
> apptainer-suid. Are you are somehow running some different types of 
> job based on singularity instead of runc?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/apptainer/apptainer/issues/611#issuecomment-1213430346>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ACDJ4FB6UGH4DUGMK3GSINLVY2OARANCNFSM56H6A6TA>.
> You are receiving this because you modified the open/close 
> state.Message ID: ***@***.***>
>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,11 @@ For older changes see the [archived Singularity change log](https://github.com/a
 
 ## Changes since last release
 
+- Fixed longstanding bug in the underlay logic when there are nested bind
+  points separated by more than one path level, for example `/var` and
+  `/var/lib/yum`, and the path didn't exist in the container image.
+  The bug only caused an error when there was a directory in the container
+  image that didn't exist on the host.
 - Improved wildcard matching in the %files directive of build definition
   files by replacing usage of sh with the mvdan.cc library.
 - Replaced checks for compatible filesystem types when using fuse-overlayfs

--- a/e2e/actions/actions.go
+++ b/e2e/actions/actions.go
@@ -2448,6 +2448,7 @@ func E2ETests(env e2e.TestEnv) testhelper.Tests {
 		"issue 5631":            c.issue5631,           // https://github.com/apptainer/singularity/issues/5631
 		"issue 5690":            c.issue5690,           // https://github.com/apptainer/singularity/issues/5690
 		"issue 6165":            c.issue6165,           // https://github.com/apptainer/singularity/issues/6165
+		"issue 619":             c.issue619,            // https://github.com/apptainer/apptainer/issues/619
 		"network":               c.actionNetwork,       // test basic networking
 		"binds":                 c.actionBinds,         // test various binds with --bind and --mount
 		"exit and signals":      c.exitSignals,         // test exit and signals propagation

--- a/e2e/actions/regressions.go
+++ b/e2e/actions/regressions.go
@@ -701,3 +701,20 @@ func (c actionTests) issue6165(t *testing.T) {
 		),
 	)
 }
+
+// Check that a nested bind separated by more than one level works with
+// underlay when the lower directory does not existing in the image but
+// the image has another directory that does not exist on the host
+func (c actionTests) issue619(t *testing.T) {
+	e2e.EnsureImage(t, c.env)
+
+	// use of user namespace to force runtime to use underlay
+	c.env.RunApptainer(
+		t,
+		e2e.WithProfile(e2e.UserNamespaceProfile),
+		e2e.WithDir(c.env.TestDir),
+		e2e.WithCommand("exec"),
+		e2e.WithArgs("--bind", "/var,/var/lib/apt", c.env.ImagePath, "true"),
+		e2e.ExpectExit(0),
+	)
+}

--- a/internal/pkg/util/fs/layout/layer/underlay/underlay_linux.go
+++ b/internal/pkg/util/fs/layout/layer/underlay/underlay_linux.go
@@ -126,6 +126,8 @@ func (u *Underlay) createLayer(rootFsPath string, system *mount.System) error {
 		return createdPath[i].len < createdPath[j].len
 	})
 
+	// Duplicate the parent directories of the new path, unless
+	//  it has already been done by another bind
 	for _, pl := range createdPath {
 		splitted := strings.Split(filepath.Dir(pl.path), string(os.PathSeparator))
 		l := len(splitted)
@@ -142,7 +144,7 @@ func (u *Underlay) createLayer(rootFsPath string, system *mount.System) error {
 				// if the directory is overridden by a bind mount we won't
 				// need to duplicate the container image directory
 				if _, ok := destinations[p]; ok {
-					continue
+					break
 				}
 				// directory not overridden, duplicate it
 				if err := u.duplicateDir(p, system, pl.path); err != nil {

